v3/src/models/data/data-set.ts

```diff
@@ -919,22 +919,74 @@ export const DataSet = V2Model.named("DataSet").props({
       },
 
       removeCases(caseIDs: string[]) {
-        caseIDs.forEach((caseID) => {
-          const index = self.getItemIndex(caseID)
+        // Remove the items last -> first, so we only have to update itemInfo once
+        const items = caseIDs.map(id => ({ id, index: self.getItemIndex(id) }))
+          .filter(info => info.index != null) as { id: string, index: number }[]
+        items.sort((a, b) => b.index - a.index)
+        const firstIndex = items[items.length - 1]?.index ?? -1
+        items.forEach(({ id: caseID, index }) => {
           if (index != null) {
             self.itemIds.splice(index, 1)
             self.attributes.forEach((attr) => {
               attr.removeValues(index)
             })
             self.selection.delete(caseID)
             self.itemInfoMap.delete(caseID)
-            for (let i = index; i < self.items.length; ++i) {
-              const itemId = self.items[i].__id__
-              const itemInfo = self.itemInfoMap.get(itemId)
-              if (itemInfo) itemInfo.index = i
-            }
           }
         })
+        if (firstIndex >= 0) {
+          for (let i = firstIndex; i < self.itemIds.length; ++i) {
+            const itemId = self.itemIds[i]
+            const itemInfo = self.itemInfoMap.get(itemId)
+            if (itemInfo) itemInfo.index = i
+          }
+        }
+      },
+
+      // If order is "first", items will be moved to the front. Otherwise, they are moved to the end.
+      moveItems(itemIds: string[], order?: string) {
```

Comment:
It would be more useful as well as more consistent with the existing API for this to take a `beforeId` for the destination, and to append if no `beforeId` is provided. I don't think it complicates the implementation much -- remove the items being moved, check the index of `beforeId` in the remaining cases, and append if it's not found.